### PR TITLE
GH-26: TokenRules unit tests.

### LINKED ISSUE #26: TokenRules unit tests.

### PR DESCRIPTION
**Pull request by pgev**
fixes #26

### CODE REVIEW

File: test/token_rules/add_global_constraint.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+            const nonOrganizationAddress = accountProvider.get();
+
+            await Utils.expectRevert(
+                tokenRules.addGlobalConstraint(
+                    constraintAddress,
+                    { from: nonOrganizationAddress },
+                ),
+                'Should revert as non-organization address is adding constraint.',
+                'Only organization is allowed to call',
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/token_rules/add_global_constraint.js]: In some cases, error.reason is coming undefined, however error.message is not. I've changed the code in a way that it fallbacks checking with error.message when error.reason is undefined. I would propose to have a ticket there to check why do we receive undefined in some rare cases.

File: test/token_rules/constructor.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const utils = require('../test_lib/utils.js');
+const { AccountProvider } = require('../test_lib/utils');
+
+const TokenRules = artifacts.require('TokenRules');
+
+contract('TokenRules::constructor', async () => {
+    contract('Negative Tests', async (accounts) => {
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/token_rules/constructor.js]: True. Added.

File: test/token_rules/register_rule.js [hunk 3]
```diff
@@ -15,118 +15,118 @@
...
+            const nonOrganizationAddress = accountProvider.get();
 
             await utils.expectRevert(
                 tokenRules.registerRule(
                     ruleName,
                     ruleAddress,
                     ruleAbi,
                     {
-                        from: nonOrganization,
+                        from: nonOrganizationAddress,
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/token_rules/register_rule.js]: I am not fully following on this point. Could you please elaborate more?

File: test/token_rules/remove_global_constraint.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                'Should revert as constraint to remove does not exist.',
+                'Constraint to remove does not exist',
+            );
+        });
+    });
+
+    contract('Events', async (accounts) => {
+        const accountProvider = new AccountProvider(accounts);
+
+        it('GlobalConstraintRemoved is emitted.', async () => {
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/token_rules/remove_global_constraint.js]: Agree! Done.

File: test/token_rules/register_rule.js [hunk 5]
```diff
@@ -15,118 +15,118 @@
...
         });
-        it('Rule with the same address.', async () => {
-            const organization = accounts[0];
-            const token = accounts[1];
+
+        it('Reverts if rule with the same address already registered.', async () => {
+            const {
+                tokenRules,
+                organizationAddress,
+            } = await TokenRulesUtils.createTokenRules(accountProvider);
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/token_rules/register_rule.js]: Done. Also, in other places.

File: test/token_rules/utils.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    return false;
+};
+
+/**
+ * Creates and returns the tuple:
+ *      (tokenRules, organizationAddress, token)
+ */
+module.exports.createTokenRules = async (accountProvider) => {
```
[the remark this answers is not in the document]
Comment (pgev) [reply test/token_rules/utils.js]: Agree. Done.

File: contracts/TokenRules.sol [hunk 7]
```diff
@@ -335,6 +331,9 @@ contract TokenRules {
     }
 
     /**
+     * @dev Function requies:
```
Comment (abhayks1) [inline contracts/TokenRules.sol]: requies => requires

File: test/token_rules/register_rule.js [hunk 8]
```diff
@@ -15,192 +15,185 @@
...
                     bRuleName,
                     bRuleAddress,
                     bRuleAbi,
-                    {
-                        from: organization,
-                    },
+                    { from: organizationAddress },
                 ),
-                'Rule with the same name can not be registered.',
+                'Should revert as a rule with the same name already registered',
```
Comment (abhayks1) [inline test/token_rules/register_rule.js]: registered => registered.

File: test/token_rules/remove_global_constraint.js [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+                'Should revert as constraint to remove does not exist.',
+                'Constraint to remove does not exist.',
+            );
+        });
+    });
+
+    contract('Events', async (accounts) => {
+        const accountProvider = new AccountProvider(accounts);
+
+        it('Emitts GlobalConstraintRemoved on removing global constraint.', async () => {
```
Comment (abhayks1) [inline test/token_rules/remove_global_constraint.js]: Emitts => Emits

**Pull request merged**